### ISSUE TITLE
InModuleScope: Using `Should -Invoke` with parameter filter inside InModuleScope fails

### ISSUE THREAD

**Issue by johlju** [pester/Pester#1813]
## General summary of the issue
When for example mocking the cmdlet `New-Object` it is important to set the parameter filter to be explicit what the mock should be triggered on. If the mock is done outside of the `InModuleScope`, and when using an `Should -Invoke` inside a `InModuleScope` using the same filter as the mock, the `Should -Invoke` fails.

## Describe your environment
```
Pester version     : 5.1.1 C:\source\SqlServerDsc\output\RequiredModules\Pester\5.1.1\Pester.psm1
PowerShell version : 7.1.0
OS version         : Microsoft Windows NT 10.0.19042.0
```

## Steps to reproduce
Running this will pass the assert on `Get-MyAlert` but fail on the assert of `New-Object`.

```powershell
BeforeAll {
    $script:moduleName = 'MyModule'

    Remove-Module -Name 'MyModule' -Force -ErrorAction 'SilentlyContinue'

    New-Module -Name 'MyModule' -ScriptBlock {
        function Get-MyAlert
        {
        }

        function New-MyAlert
        {
            Get-MyAlert

            $null = New-Object -TypeName Microsoft.SqlServer.Management.Smo.Agent.Alert
        }
    } | Import-Module

    $PSDefaultParameterValues = @{
        'InModuleScope:ModuleName' = $script:moduleName
    }
}

Describe 'InModuleScope' {
    BeforeAll {
        Mock -CommandName Get-MyAlert
        Mock -CommandName New-Object -MockWith {
            return 'anything'
        } -ParameterFilter {
            $TypeName -eq 'Microsoft.SqlServer.Management.Smo.Agent.Alert'
        }
    }

    It 'Should call the mock' {
        InModuleScope -ScriptBlock {
            { New-MyAlert } | Should -Not -Throw

            Should -Invoke -CommandName Get-MyAlert -Exactly -Times 1 -Scope It

            Should -Invoke -CommandName New-Object -ParameterFilter {
                $TypeName -eq 'Microsoft.SqlServer.Management.Smo.Agent.Alert'
            } -Exactly -Times 1 -Scope It
        }
    }
}
```

## Expected Behavior
Expected the assert (`Should -Invoke`) for `New-Object` to pass since the assert for `Get-MyAlert` passed.

This might be by design, so this can just be a documentation issue. A rule might be that if the mock is outside of the `InModuleScope`, the assert should be as well.

## Current Behavior
The assert (`Should -Invoke`) using parameter filter inside a `InModuleScope`  fails.

## Possible Solution? (optional)

Moving the assert with the parameter filter outside of the `InModuleScope` makes the assert to pass.

```powershell
BeforeAll {
    $script:moduleName = 'MyModule'

    Remove-Module -Name 'MyModule' -Force -ErrorAction 'SilentlyContinue'

    New-Module -Name 'MyModule' -ScriptBlock {
        function Get-MyAlert
        {
        }

        function New-MyAlert
        {
            Get-MyAlert

            $null = New-Object -TypeName Microsoft.SqlServer.Management.Smo.Agent.Alert
        }
    } | Import-Module

    $PSDefaultParameterValues = @{
        'InModuleScope:ModuleName' = $script:moduleName
    }
}

Describe 'InModuleScope' {
    BeforeAll {
        Mock -CommandName Get-MyAlert
        Mock -CommandName New-Object -MockWith {
            return 'anything'
        } -ParameterFilter {
            $TypeName -eq 'Microsoft.SqlServer.Management.Smo.Agent.Alert'
        }
    }

    It 'Should call the mock' {
        InModuleScope -ScriptBlock {
            { New-MyAlert } | Should -Not -Throw

            Should -Invoke -CommandName Get-MyAlert -Exactly -Times 1 -Scope It
        }

        Should -Invoke -CommandName New-Object -ParameterFilter {
            $TypeName -eq 'Microsoft.SqlServer.Management.Smo.Agent.Alert'
        } -Exactly -Times 1 -Scope It
    }
}
```

**Comment by fflaten** [comment]
Thanks for another detailed report! 👍 

There's a few things here. As mentioned in the other issue, `-ModuleName` parameter in `Mock` and `Should -Invoke` is used to specify which state the mock should be available to call. So to limit your `New-Object` mock to only affect calls done from a place (ex. a function) inside 'MyModule', you need to specify `-ModuleName 'MyModule'` both when creating the mocks and when asserting if they were called **unless** those functions are used inside `InModuleScope`. 

The fact that you're "possible solution" script works is actually a bug that was fixed in 5.2.0. The `New-Object` should only be visible in the script scope since you didn't specify `-ModuleName MyModule` when defining it, but since it originates from another module, the mock "leaks". This behavior was changed in 5.2.0 to now be more strict and require `InModuleScope` or  `Mock -ModuleName ...`. Try:

```powershell
BeforeAll {
    $script:moduleName = 'MyModule'

    Remove-Module -Name 'MyModule' -Force -ErrorAction 'SilentlyContinue'

    New-Module -Name 'MyModule' -ScriptBlock {
        function Get-MyAlert {
            write-warning "real function called!"
        }

        function New-MyAlert {
            Get-MyAlert

            $null = New-Object -TypeName Microsoft.SqlServer.Management.Smo.Agent.Alert
        }
    } | Import-Module

    $PSDefaultParameterValues = @{
        'InModuleScope:ModuleName' = $script:moduleName
    }
}

Describe 'InModuleScope' {
    BeforeAll {
        Mock -CommandName Get-MyAlert -ModuleName $script:moduleName
        Mock -CommandName New-Object -ModuleName $script:moduleName -MockWith {
            return 'anything'
        } -ParameterFilter {
            $TypeName -eq 'Microsoft.SqlServer.Management.Smo.Agent.Alert'
        }
    }

    It 'Should call the mock' {
        InModuleScope -ScriptBlock {
            { New-MyAlert } | Should -Not -Throw
        }

        Should -Invoke -CommandName Get-MyAlert -ModuleName $script:moduleName -Exactly -Times 1 -Scope It

        Should -Invoke -CommandName New-Object -ModuleName $script:moduleName -ParameterFilter {
            $TypeName -eq 'Microsoft.SqlServer.Management.Smo.Agent.Alert'
        } -Exactly -Times 1 -Scope It
    }
}
```

There is one remaining issue with my solution above though. I've moved the `Should -Invoke` calls outside your `InModuleScope` block to minimize usage of `InModuleScope ` (see tip at the end of [this page](https://pester.dev/docs/usage/modules)). The example below should also have worked AFAIK, but it currnetly don't. IIRC the fixes made in 5.2.0 were only done for `Mock`, so there is some remaining work to fix `Should -Invoke` inside `InModuleScope`. Or am I missing something here @nohwnd  ? 🙂 

```powershell
    It 'Should call the mock' {
        InModuleScope -ScriptBlock {
            { New-MyAlert } | Should -Not -Throw

            Should -Invoke -CommandName Get-MyAlert -Exactly -Times 1 -Scope It

            Should -Invoke -CommandName New-Object -ParameterFilter {
                $TypeName -eq 'Microsoft.SqlServer.Management.Smo.Agent.Alert'
            } -Exactly -Times 1 -Scope It
        }
    }
```

**Comment by nohwnd** [comment]
@fflaten your last example does not work because the internal wrapper that sets the variables in Test-ParameterFilter is bound to incorrect session state. So the variable is set, it is shown in the detailed log, but then the actual scriptblock does not have it, because we've set it to script scope and the parameter filter scriptblock is running in module. Change filter to $true and it works.

The erro is that I am setting it externally to avoid doing reflection on the scriptblock (which would be the most correct solution, as the current one can still be broken in some ways), and in this case the caller sessionstate of the mock hook is used incorrectly. Instead the $SessionState should be used, which we get few levels above from $PSCmdlet.SessionState (caller session state). 

https://github.com/pester/Pester/blob/v5.0/src/functions/Mock.ps1#L397

Just make that line say = $SessionState and it will correct. If you need more info or help debugging to be able to see this problem let me know.

**Comment by fflaten** [comment]
Closing as this should be fixed by specifying `-ModuleName MyModule` when creating the mocks, see comments above.
The last part about being able to run `Should -Invoke` inside `InModuleScope` was fixed in #1954 and will be available in the next release.
